File: src/netconsole/models/traffic_test.py
from __future__ import annotations

from dataclasses import dataclass, field, replace
from enum import StrEnum
from typing import Any

from netconsole.models.task_snapshot import utc_now_iso
from netconsole.models.task_state import TaskState


class TrafficTestType(StrEnum):
    IPERF_SERVER = "IPERF_SERVER"
    IPERF_CLIENT = "IPERF_CLIENT"
    HIGH_FREQUENCY_PING = "HIGH_FREQUENCY_PING"


class ExecutionTargetKind(StrEnum):
    LOCAL = "LOCAL"
    AGENT = "AGENT"


class TrafficEventType(StrEnum):
    STATE = "state"
    STDOUT = "stdout"
    STDERR = "stderr"
    SAMPLE = "sample"
    SUMMARY = "summary"
    ERROR = "error"
    SYSTEM = "system"


class TrafficSyncState(StrEnum):
    ACTIVE = "ACTIVE"
    STALE = "STALE"
    CREDENTIAL_REQUIRED = "CREDENTIAL_REQUIRED"
    AGENT_OFFLINE = "AGENT_OFFLINE"
    COMPLETED = "COMPLETED"
    ERROR = "ERROR"


@dataclass(frozen=True)
class ExecutionTargetDTO:
    kind: ExecutionTargetKind
    agent_id: str = ""
    display_name: str = ""

    def __post_init__(self) -> None:
        if self.kind is ExecutionTargetKind.LOCAL and self.agent_id:
            raise ValueError("LOCAL execution target cannot contain agent_id")
        if self.kind is ExecutionTargetKind.AGENT and not self.agent_id.strip():
            raise ValueError("AGENT execution target requires agent_id")


@dataclass(frozen=True)
class HighFrequencyPingConfig:
    targets: tuple[str, ...]
    interval_ms: int = 100
    timeout_ms: int = 100
    packet_size: int = 64
    count: int = 20
    continuous: bool = False
    source_address: str = ""

    def normalized(self) -> HighFrequencyPingConfig:
        targets = tuple(str(value).strip() for value in self.targets)
        if not targets or any(not value for value in targets):
            raise ValueError("at least one ping target is required")
        if len(targets) > 64:
            raise ValueError("ping targets cannot exceed 64")
        if len(set(targets)) != len(targets):
            raise ValueError("ping targets cannot contain duplicates")
        interval_ms = int(self.interval_ms)
        timeout_ms = int(self.timeout_ms)
        packet_size = int(self.packet_size)
        count = int(self.count)
        if not 1 <= interval_ms <= 60_000:
            raise ValueError("interval_ms must be between 1 and 60000")
        if not 1 <= timeout_ms <= 60_000:
            raise ValueError("timeout_ms must be between 1 and 60000")
        if not 1 <= packet_size <= 65_507:
            raise ValueError("packet_size must be between 1 and 65507")
        if self.continuous:
            if count != 0:
                raise ValueError("continuous ping requires count=0")
        elif not 1 <= count <= 1_000_000:
            raise ValueError("finite ping count must be between 1 and 1000000")
        return replace(
            self,
            targets=targets,
            interval_ms=interval_ms,
            timeout_ms=timeout_ms,
            packet_size=packet_size,
            count=count,
            source_address=str(self.source_address or "").strip(),
        )

    def to_dict(self) -> dict[str, Any]:
        value = self.normalized()
        return {
            "targets": list(value.targets),
            "interval_ms": value.interval_ms,
            "timeout_ms": value.timeout_ms,
            "packet_size": value.packet_size,
            "count": value.count,
            "continuous": value.continuous,
            "source_address": value.source_address,
        }


@dataclass(frozen=True)
class TrafficRun:
    traffic_run_id: str
    test_type: TrafficTestType
    role: str
    executor_kind: ExecutionTargetKind
    normalized_config: dict[str, Any]
    status: TaskState
    created_at: str
    updated_at: str
    controller_task_id: str = ""
    agent_id: str = ""
    started_at: str = ""
    finished_at: str = ""
    summary: dict[str, Any] = field(default_factory=dict)
    error_code: str = ""
    error_message: str = ""
    raw_reference: str = ""
    result_reference: str = ""
    local_iperf_run_id: str = ""
    retry_of_traffic_run_id: str = ""
    parent_task_id: str = ""
    correlation_id: str = ""
    last_event_sequence: int = 0
    sync_state: TrafficSyncState = TrafficSyncState.ACTIVE


@dataclass(frozen=True)
class AgentTaskMapping:
    traffic_run_id: str
    controller_task_id: str
    agent_id: str
    agent_task_id: str
    agent_task_type: str
    last_remote_sequence: int = 0
    last_remote_status: str = "created"
    last_polled_at: str = ""
    sync_state: TrafficSyncState = TrafficSyncState.ACTIVE
    sync_error_code: str = ""
    sync_error_message: str = ""
    created_at: str = ""
    updated_at: str = ""


@dataclass(frozen=True)
class TrafficPingSample:
    traffic_run_id: str
    sequence: int
    timestamp: str
    target: str
    probe_sequence: int | None
    ok: bool
    rtt_ms: float | None
    timeout: bool = False
    packet_size: int | None = None
    error_code: str = ""
    error_message: str = ""


@dataclass(frozen=True)
class TrafficEvent:
    traffic_run_id: str
    controller_task_id: str
    source: str
    type: TrafficEventType
    payload: dict[str, Any] = field(default_factory=dict)
    timestamp: str = field(default_factory=utc_now_iso)
    sequence: int = 0
    remote_sequence: int | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "sequence": int(self.sequence),
            "timestamp": self.timestamp,
            "traffic_run_id": self.traffic_run_id,
            "controller_task_id": self.controller_task_id,
            "source": self.source,
            "type": self.type.value,
            "payload": dict(self.payload),
            "remote_sequence": self.remote_sequence,
        }

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> TrafficEvent:
        remote_sequence = value.get("remote_sequence")
        return cls(
            sequence=int(value.get("sequence") or 0),
            timestamp=str(value.get("timestamp") or ""),
            traffic_run_id=str(value.get("traffic_run_id") or ""),
            controller_task_id=str(value.get("controller_task_id") or ""),
            source=str(value.get("source") or ""),
            type=TrafficEventType(str(value.get("type") or TrafficEventType.SYSTEM.value)),
            payload=dict(value.get("payload") or {}),
            remote_sequence=int(remote_sequence) if remote_sequence is not None else None,
        )

File: apps/web/src/stores/traffic.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { createPinia, setActivePinia } from 'pinia'

import { useTrafficStore } from './traffic'
import {
  getTrafficRun,
  getTrafficSummary,
  listTrafficEvents,
  listTrafficExecutionTargets,
  listTrafficPingSamples,
  listTrafficRuns,
  startFping,
} from '../api/traffic'
import type { TrafficEvent, TrafficExecutionTarget, TrafficPingSample, TrafficRun } from '../types/traffic'

vi.mock('../api/traffic', () => ({
  listTrafficExecutionTargets: vi.fn(),
  startIperfServer: vi.fn(),
  startIperfClient: vi.fn(),
  startFping: vi.fn(),
  listTrafficRuns: vi.fn(),
  getTrafficRun: vi.fn(),
  getTrafficSummary: vi.fn(),
  listTrafficEvents: vi.fn(),
  listTrafficPingSamples: vi.fn(),
  cancelTrafficRun: vi.fn(),
  retryTrafficRun: vi.fn(),
}))

const localTarget: TrafficExecutionTarget = {
  kind: 'LOCAL',
  id: 'LOCAL',
  display_name: '本机',
  available: true,
  unavailable_reason: '',
  agent_id: '',
  status: '',
  platform: '',
  architecture: '',
  version: '',
  capabilities: { fping: true, iperf_server: true, iperf_client: true },
}

const run: TrafficRun = {
  id: 'run-1',
  traffic_run_id: 'run-1',
  controller_task_id: 'task-1',
  test_type: 'HIGH_FREQUENCY_PING',
  role: 'ping',
  executor_kind: 'LOCAL',
  agent_id: '',
  normalized_config: { targets: ['192.0.2.1'] },
  status: 'RUNNING',
  created_at: '2026-07-13T00:00:00Z',
  started_at: '2026-07-13T00:00:00Z',
  finished_at: '',
  updated_at: '2026-07-13T00:00:01Z',
  summary: { sent: 1 },
  error_code: '',
  error_message: '',
  raw_reference: '',
  result_reference: '',
  retry_of_traffic_run_id: '',
  parent_task_id: '',
  correlation_id: '',
  last_event_sequence: 1,
  sync_state: 'ACTIVE',
  cancellable: true,
}

const event: TrafficEvent = {
  sequence: 1,
  timestamp: '2026-07-13T00:00:01Z',
  traffic_run_id: 'run-1',
  controller_task_id: 'task-1',
  source: 'local',
  type: 'sample',
  payload: { rtt_ms: 1.2 },
  remote_sequence: null,
}

const sample: TrafficPingSample = {
  traffic_run_id: 'run-1',
  sequence: 1,
  timestamp: '2026-07-13T00:00:01Z',
  target: '192.0.2.1',
  probe_sequence: 1,
  ok: true,
  rtt_ms: 1.2,
  timeout: false,
  packet_size: 64,
  error_code: '',
  error_message: '',
}

class FakeWebSocket {
  static OPEN = 1
  static instances: FakeWebSocket[] = []
  readyState = 0
  onopen: (() => void) | null = null
  onmessage: ((event: { data: string }) => void) | null = null
  onclose: (() => void) | null = null
  onerror: (() => void) | null = null

  constructor(public url: string) { FakeWebSocket.instances.push(this) }
  open(): void { this.readyState = 1; this.onopen?.() }
  receive(value: unknown): void { this.onmessage?.({ data: JSON.stringify(value) }) }
  close(): void { this.readyState = 3; this.onclose?.() }
}

describe('Traffic store', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    vi.mocked(listTrafficExecutionTargets).mockReset().mockResolvedValue([localTarget])
    vi.mocked(listTrafficRuns).mockReset().mockResolvedValue([run])
    vi.mocked(getTrafficRun).mockReset().mockResolvedValue(run)
    vi.mocked(getTrafficSummary).mockReset().mockResolvedValue({ traffic_run_id: 'run-1', updated_at: run.updated_at, summary: { sent: 1 } })
    vi.mocked(listTrafficEvents).mockReset().mockResolvedValue([event])
    vi.mocked(listTrafficPingSamples).mockReset().mockResolvedValue([sample])
    vi.mocked(startFping).mockReset().mockResolvedValue({ run })
    FakeWebSocket.instances = []
    vi.stubGlobal('WebSocket', FakeWebSocket)
    vi.stubGlobal('window', {
      location: { protocol: 'http:', host: '127.0.0.1:8000' },
      setTimeout,
      clearTimeout,
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
    })
  })

  it('loads targets, runs and starts fping through the traffic API', async () => {
    const store = useTrafficStore()
    await store.refreshTargets()
    await store.refreshRuns()
    expect(store.targets).toEqual([localTarget])
    expect(store.runningCount).toBe(1)

    await store.createFping({
      execution_target: { kind: 'LOCAL' },
      targets: ['192.0.2.1'],
      interval_ms: 100,
      timeout_ms: 100,
      packet_size: 64,
      count: 3,
      continuous: false,
    })

    expect(startFping).toHaveBeenCalledOnce()
    expect(store.selected?.traffic_run_id).toBe('run-1')
  })

  it('connects with REST cursors and merges dedicated incremental samples', async () => {
    const store = useTrafficStore()
    await store.selectRun('run-1')
    expect(FakeWebSocket.instances[0].url).toBe('ws://127.0.0.1:8000/ws/traffic/run-1?after_event=1&after_sample=1')
    FakeWebSocket.instances[0].open()
    FakeWebSocket.instances[0].receive({ type: 'samples', samples: [{ ...sample, sequence: 2, rtt_ms: 2.4 }] })
    FakeWebSocket.instances[0].receive({
      type: 'events',
      events: [{ ...event, sequence: 2, type: 'state', payload: { state: 'COMPLETED' } }],
    })
    expect(store.socketConnected).toBe(true)
    expect(store.samples.map((item) => item.sequence)).toEqual([1, 2])
    expect(store.selected?.status).toBe('COMPLETED')
    store.disconnectSocket()
  })
})
